File: mobile/modules/expo-audio-segment-stitcher/android/src/main/java/com/ehllo/app/audiosegmentstitcher/ExpoAudioSegmentStitcherModule.kt
package com.ehllo.app.audiosegmentstitcher

import android.media.MediaCodec
import android.media.MediaExtractor
import android.media.MediaFormat
import android.media.MediaMuxer
import android.net.Uri
import java.io.File
import java.nio.ByteBuffer
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoAudioSegmentStitcherModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoAudioSegmentStitcher")

    // Combines recording segments (each its own finalized .m4a — expo-audio
    // can't reopen and append to an already-stopped file, e.g. after the
    // interruption watchdog forces a new take) into one continuous .m4a via
    // MediaExtractor/MediaMuxer, so playback/upload always sees a single file.
    AsyncFunction("concatenateSegments") { uris: List<String> ->
      concatenate(uris)
    }
  }

  private fun resolvePath(uriString: String): String {
    val uri = Uri.parse(uriString)
    return uri.path?.takeIf { uri.scheme == "file" } ?: uriString
  }

  private fun concatenate(uris: List<String>): String {
    if (uris.isEmpty()) throw Exception("No audio segments were provided.")
    if (uris.size == 1) return uris[0]

    val outputFile = File.createTempFile("ehllo-stitched-", ".m4a")
    val muxer = MediaMuxer(outputFile.absolutePath, MediaMuxer.OutputFormat.MUXER_OUTPUT_MPEG_4)

    var audioTrackIndex = -1
    var muxerStarted = false
    var timeOffsetUs = 0L
    val bufferInfo = MediaCodec.BufferInfo()
    val buffer = ByteBuffer.allocate(1 * 1024 * 1024)

    try {
      for (uriString in uris) {
        val path = resolvePath(uriString)
        val extractor = MediaExtractor()
        extractor.setDataSource(path)

        var trackIndex = -1
        var format: MediaFormat? = null
        for (i in 0 until extractor.trackCount) {
          val candidate = extractor.getTrackFormat(i)
          val mime = candidate.getString(MediaFormat.KEY_MIME) ?: ""
          if (mime.startsWith("audio/")) {
            trackIndex = i
            format = candidate
            break
          }
        }
        if (trackIndex == -1 || format == null) {
          extractor.release()
          throw Exception("Segment has no audio track: $uriString")
        }
        extractor.selectTrack(trackIndex)

        if (!muxerStarted) {
          audioTrackIndex = muxer.addTrack(format)
          muxer.start()
          muxerStarted = true
        }

        var maxPresentationTimeUs = 0L
        while (true) {
          buffer.clear()
          val sampleSize = extractor.readSampleData(buffer, 0)
          if (sampleSize < 0) break
          val presentationTimeUs = extractor.sampleTime
          bufferInfo.offset = 0
          bufferInfo.size = sampleSize
          bufferInfo.presentationTimeUs = presentationTimeUs + timeOffsetUs
          bufferInfo.flags = extractor.sampleFlags
          muxer.writeSampleData(audioTrackIndex, buffer, bufferInfo)
          if (presentationTimeUs > maxPresentationTimeUs) maxPresentationTimeUs = presentationTimeUs
          extractor.advance()
        }
        timeOffsetUs += maxPresentationTimeUs
        extractor.release()
      }
    } finally {
      try {
        muxer.stop()
      } catch (_: Exception) {
        // Nothing more we can do if the muxer never got a valid track/samples.
      }
      muxer.release()
    }

    return Uri.fromFile(outputFile).toString()
  }
}

File: mobile/modules/expo-audio-segment-stitcher/ios/ExpoAudioSegmentStitcherModule.swift
import ExpoModulesCore
import AVFoundation

enum AudioSegmentStitchError: Error, LocalizedError {
  case noSegments
  case trackCreationFailed
  case invalidUri(String)
  case noAudioTrack(String)
  case exportSessionFailed
  case exportFailed(String)

  var errorDescription: String? {
    switch self {
    case .noSegments:
      return "No audio segments were provided."
    case .trackCreationFailed:
      return "Could not create a composition track for stitching."
    case .invalidUri(let uri):
      return "Invalid audio segment URI: \(uri)"
    case .noAudioTrack(let uri):
      return "Segment has no audio track: \(uri)"
    case .exportSessionFailed:
      return "Could not create an export session for the stitched recording."
    case .exportFailed(let message):
      return "Export failed: \(message)"
    }
  }
}

public class ExpoAudioSegmentStitcherModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoAudioSegmentStitcher")

    // Combines recording segments (each its own finalized .m4a - expo-audio
    // can't reopen and append to an already-stopped file, e.g. after the
    // interruption watchdog forces a new take) into one continuous .m4a via
    // AVFoundation composition, so playback/upload always sees a single file.
    AsyncFunction("concatenateSegments") { (uris: [String]) -> String in
      try self.concatenate(uris: uris)
    }
  }

  private func concatenate(uris: [String]) throws -> String {
    guard !uris.isEmpty else { throw AudioSegmentStitchError.noSegments }
    guard uris.count > 1 else { return uris[0] }

    let composition = AVMutableComposition()
    guard let audioTrack = composition.addMutableTrack(
      withMediaType: .audio,
      preferredTrackID: kCMPersistentTrackID_Invalid
    ) else {
      throw AudioSegmentStitchError.trackCreationFailed
    }

    var insertTime = CMTime.zero
    for uriString in uris {
      guard let url = fileURL(from: uriString) else {
        throw AudioSegmentStitchError.invalidUri(uriString)
      }
      let asset = AVURLAsset(url: url)
      guard let sourceTrack = asset.tracks(withMediaType: .audio).first else {
        throw AudioSegmentStitchError.noAudioTrack(uriString)
      }
      let range = CMTimeRange(start: .zero, duration: asset.duration)
      try audioTrack.insertTimeRange(range, of: sourceTrack, at: insertTime)
      insertTime = CMTimeAdd(insertTime, asset.duration)
    }

    let outputURL = FileManager.default.temporaryDirectory
      .appendingPathComponent("aftermeet-stitched-\(UUID().uuidString).m4a")

    guard let exportSession = AVAssetExportSession(
      asset: composition,
      presetName: AVAssetExportPresetAppleM4A
    ) else {
      throw AudioSegmentStitchError.exportSessionFailed
    }
    exportSession.outputURL = outputURL
    exportSession.outputFileType = .m4a

    let semaphore = DispatchSemaphore(value: 0)
    var exportError: Error?
    exportSession.exportAsynchronously {
      if exportSession.status != .completed {
        exportError = exportSession.error ?? AudioSegmentStitchError.exportFailed("unknown")
      }
      semaphore.signal()
    }
    semaphore.wait()

    if let exportError {
      throw AudioSegmentStitchError.exportFailed(exportError.localizedDescription)
    }

    return outputURL.absoluteString
  }

  private func fileURL(from uriString: String) -> URL? {
    if let url = URL(string: uriString), url.scheme != nil {
      return url
    }
    return URL(fileURLWithPath: uriString)
  }
}
